### PR TITLE
feat(filer webhook)

### PR DESCRIPTION
**Pull request by Jose-Matsuda**
https://jirab.statcan.ca/browse/BTIS-285

The framework of this webhook was taken from [this example](https://github.com/morvencao/kube-sidecar-injector) and will inject containers necessary for our filers to be mounted to a notebook. See the README for more information on gotchas

**IMPORTANT: Will have to persist the settings in argocd.** Essentially it is everything in the `deploy` folder

### CODE REVIEW

File: ex-pod.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: standalone-test-pod
+  #namespace: jose-matsuda
+  namespace: mathis-marcotte
```
Comment (mathis-marcotte) [inline ex-pod.yaml:6]: This file wont get commited, right?

File: deploy/deployment.yaml [hunk 2]
```diff
@@ -1,27 +1,30 @@
...
-          image: sidecar-injector
+        - name: filer-sidecar-injector
+          image: artifactory.cloud.statcan.ca/das-aaw-docker/filer-sidecar:latest
           imagePullPolicy: Always
           args:
-          - -service-name=sidecar-injector
-          - -sidecar-config-file=/etc/webhook/config/sidecarconfig.yaml
+          - -service-name=filer-sidecar-injector
+          - -sidecar-config-file=/etc/webhook/config/sidecarconfig.json
+          #- -sidecar-config-file=/etc/webhook/config/sidecarconfig.yaml
```
Comment (mathis-marcotte) [inline deploy/deployment.yaml:27]: maybe remove this commented line

File: build.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+az acr login -n k8scc01covidacr
```
Comment (mathis-marcotte) [inline build.sh:1]: Is this file getting commited?

File: cmd/main.go [hunk 4]
```diff
@@ -37,8 +37,9 @@ func init() {
 func main() {
 	// init command flags
 	flag.IntVar(&port, "port", 8443, "Webhook server port.")
-	flag.StringVar(&webhookServiceName, "service-name", "sidecar-injector", "Webhook service name.")
-	flag.StringVar(&sidecarConfigFile, "sidecar-config-file", "/etc/webhook/config/sidecarconfig.yaml", "Sidecar injector configuration file.")
+	flag.StringVar(&webhookServiceName, "service-name", "filer-sidecar-injector", "Webhook service name.")
+	//flag.StringVar(&sidecarConfigFile, "sidecar-config-file", "/etc/webhook/config/sidecarconfig.yaml", "Sidecar injector configuration file.")
```
Comment (mathis-marcotte) [inline cmd/main.go:41]: maybe remove this commented out line

File: cmd/webhook.go [hunk 5]
```diff
@@ -169,14 +168,86 @@ func updateAnnotation(target map[string]string, added map[string]string) (patch
...
+				}
+			}
+		}
+	}
+	return patch
+}
+
 // create mutation patch for resoures
-func createPatch(pod *corev1.Pod, sidecarConfig *Config, annotations map[string]string) ([]byte, error) {
+func createPatch(pod *corev1.Pod, oldSidecarConfig *Config, annotations map[string]string) ([]byte, error) {
```
Comment (mathis-marcotte) [inline cmd/webhook.go:203]: Maybe a better name for oldSidecarConfig would be something like "sidecarConfigTemplate"?

File: cmd/webhook.go [hunk 6]
```diff
@@ -169,14 +168,86 @@ func updateAnnotation(target map[string]string, added map[string]string) (patch
...
+		panic(err.Error())
+	}
+	secretList, _ := clientset.CoreV1().Secrets(pod.Namespace).List(context.Background(), metav1.ListOptions{})
+	isFirst := true
+
+	for sec := range secretList.Items {
+		// check for secrets having filer-conn-secret
+		if strings.Contains(secretList.Items[sec].Name, "filer-conn-secret") {
+			// Should deep copy because things change blah
+			sidecarConfig1, _ := deepcopy.Anything(oldSidecarConfig)
```
Comment (mathis-marcotte) [inline cmd/webhook.go:223]: maybe a better name than sidecarConfig1 would be "tempSidecarConfig"?

**Review by mathis-marcotte — CHANGES_REQUESTED** [review]

File: ex-pod.yaml (hunk 1 above)
Comment (Jose-Matsuda) [reply ex-pod.yaml:6]: I mainly kept it for easy testing of the webhook without having to start up a pod, I would like to keep it tbh. Maybe I will just have it documented somewhere

File: build.sh (hunk 3 above)
Comment (Jose-Matsuda) [reply build.sh:1]: Had it for easy testing purposes, can remove it

Comment (mathis-marcotte) [reply build.sh]: maybe just add this script in the readme? just instead of the actual file

**Review by mathis-marcotte — APPROVED** [review]
lgtm

**Pull request merged**